### PR TITLE
[hwasan] Fix and re-enable deep-recursion.c

### PR DESCRIPTION
**Pull request by thurstond**
deep-recursion.c was disabled
(https://github.com/llvm/llvm-project/commit/c007e0f66ee3f96467fd12f6200218fb4c38c2c9)
because the test may get unlucky and end up with a zero-tagged variable, leading to a false negative (https://github.com/llvm/llvm-project/issues/69221).

This patch re-enables the test and adds a workaround: it checks
if the variable is zero-tagged, and if so, it will instead use the
neighboring variable, which must have a different (hence non-zero)
tag.

Fixing the stack allocation tagging is left as an exercise for the
reader. It is non-trivial because, even if the stackTagBase is
non-zero, tags for subsequent allocations in the stack frame may wrap
around to zero; working around this would require adding multiple
instructions to each alloca.

### CODE REVIEW

**Comment by thurstond** [comment]
The latest PR update is pulling in unrelated commits. Setting this PR to draft until I sort it out.

File: compiler-rt/test/hwasan/TestCases/deep-recursion.c [hunk 1]
```diff
@@ -29,7 +26,23 @@ void USE(void *x) { // pretend_to_do_something(void *x)
...
+__attribute__((noinline)) void OOB() {
+  int x[4];
+  int y[4];
+
+  // Tags for stack-allocated variables can occasionally be zero, resulting in
+  // a false negative for this test. This is not easy to fix, hence we work
+  // around it: if the tag is zero, we use the neighboring variable instead,
+  // which must have a different (hence non-zero) tag.
+  // This tag check assumes aarch64.
+  if(((unsigned long)&x) >> 56 == 0) {
```
Comment (ilovepi) [inline compiler-rt/test/hwasan/TestCases/deep-recursion.c]: Would this be better as `uintptr_t`?

**Review by ilovepi — APPROVED** [review]
LGTM once you've sorted out the history. I left a small question inline, but it's more of a nit question than something to block on.

File: compiler-rt/test/hwasan/TestCases/deep-recursion.c (hunk 1 above)
Comment (thurstond) [reply compiler-rt/test/hwasan/TestCases/deep-recursion.c]: Good point, fixed!

**Comment by ilovepi** [comment]
I think this is still marked as a draft. I'm not sure if that was intentional or not.

**Comment by thurstond** [comment]
> I think this is still marked as a draft. I'm not sure if that was intentional or not.

Oops, I forgot to explicitly toggle it. Thanks for pointing it out!

**Review by vitalybuka — APPROVED** [review]

**Comment by vitalybuka** [comment]
I will land so we can see results by tomorrow,.

**Comment by thurstond** [comment]
> I will land so we can see results by tomorrow,.

Thank you Vitaly!

**Pull request merged**